flang/lib/Semantics/resolve-directives.cpp

```diff
@@ -2540,6 +2541,32 @@ void ResolveOmpTopLevelParts(
   });
 }
 
+static bool IsSymbolInCommonBlock(const Symbol &symbol) {
+  // If there are many symbols in common blocks, going through them all can be
+  // slow. A possible optimization is to add an OmpInCommonBlock flag to
+  // Symbol, to make it possible to quickly test if a given symbol is in a
+  // common block.
+  for (const auto &cb : symbol.owner().commonBlocks()) {
+    if (IsCommonBlockContaining(cb.second.get(), symbol)) {
+      return true;
+    }
+  }
+  return false;
+}
+
+static bool IsSymbolPredeterminedAsPrivate(const Symbol &symbol) {
```

Comment:
What about loop indices?

Reply:
Loop indices are handled by other parts of the OmpAttributeVisitor, that should run before this check and set the OmpPrivate flag in the corresponding symbols. However, I'm seeing some strange behaviors with the copyprivate check in some cases. I'll investigate it further and add some more tests.